source/Calamari/TargetDiscoveryBehaviour.cs

```diff
@@ -0,0 +1,150 @@
+{
+    public class TargetDiscoveryBehaviour : IDeployBehaviour
+    {
+        private ILog Log { get; }
+
+        public TargetDiscoveryBehaviour(ILog log)
+        {
+            Log = log;
+        }
+
+        public bool IsEnabled(RunningDeployment context) => true;
+
+        public async Task Execute(RunningDeployment runningDeployment)
+        {
+            await Task.CompletedTask;
+            var targetDiscoveryContext = GetTargetDiscoveryContext(runningDeployment.Variables);
+            if (targetDiscoveryContext == null)
+            {
+                Log.Warn("Aborting target discovery.");
+                return;
+            }
+
+            var account = targetDiscoveryContext.Authentication.AccountDetails;
+            var azureClient = account.CreateAzureClient();
+
+            try
+            {
+                var webApps = azureClient.WebApps.ListWebAppBasic();
+                foreach (var webApp in webApps)
+                {
+                    var tags = GetTags(webApp);
+                    var matchResult = targetDiscoveryContext.Scope.Match(tags);
+                    if (matchResult.IsSuccess)
+                    {
+                        Log.Info($"Discovered matching web app: {webApp.Name}");
+                        WriteTargetCreationServiceMessage(
+                            webApp, targetDiscoveryContext, matchResult, runningDeployment.Variables);
+                    }
+                }
+
+                if (!webApps.Any())
+                {
+                    Log.Info("No matching web apps found.");
+                }
+            }
+            catch (Exception ex)
+            {
+                Log.Warn($"Error connecting to Azure to look for web apps:");
+                Log.Warn(ex.Message);
+                Log.Warn("Aborting target discovery.");
+            }
+        }
+
+        private TargetTags GetTags(IWebAppBasic webApp)
+        {
+            webApp.Tags.TryGetValue(TargetTags.EnvironmentTagName, out string? environment);
+            webApp.Tags.TryGetValue(TargetTags.RoleTagName, out string? role);
+            webApp.Tags.TryGetValue(TargetTags.ProjectTagName, out string? project);
+            webApp.Tags.TryGetValue(TargetTags.SpaceTagName, out string? space);
+            webApp.Tags.TryGetValue(TargetTags.TenantTagName, out string? tenant);
+            return new TargetTags(
+                environment: environment,
+                role: role,
+                project: project,
+                space: space,
+                tenant: tenant);
+        }
+
+        private void WriteTargetCreationServiceMessage(
+            IWebAppBasic webApp,
+            TargetDiscoveryContext<AccountAuthenticationDetails<ServicePrincipalAccount>> context,
+            TargetMatchResult matchResult,
+            IVariables variables)
+        {
+            // TODO: extend target discovery context to include account ID and worker pool ID directly?
```

Comment:
Is this comment still applicable about account and worker pool ids?

Reply:
No! Removed.